src/packageurl/contrib/url2purl.py

```diff
@@ -262,17 +280,183 @@ def build_nuget_purl(uri):
+            version = segments[4]
+
+        return PackageURL(
+            type="github", namespace=namespace, name=name, version=version
+        )
+
+    return
+
+
+github_codeload_pattern = (
+    r"https?://codeload.github.com/"
+    r"(?P<namespace>.+)/(?P<name>.+)/(zip|tar.gz|tar.bz2|.tgz)/v(?P<version>.+)$"
+)
+
+
+@purl_router.route(github_codeload_pattern)
+def build_github_purl(uri):
+    """
+    example :  https://codeload.github.com/nexB/scancode-toolkit/tar.gz/v3.1.1
+    """
+    return purl_from_pattern("github", github_codeload_pattern, uri)
+
+
+@purl_router.route("https?://github\\.com/.*")
+def build_github_purl(uri):
+    """
+    Take `uri` as input 
+    For example:
+    https://github.com/nexB/scancode-toolkit/tree/develop/plugins/scancode-ctags-macosx_10_9_intel
+    """
+    regex = r"https?://github.com/(?P<namespace>.+)/(?P<name>.+)/archive/v(?P<version>.+).(zip|tar.gz|tar.bz2|.tgz)"
+    matches = re.search(regex, uri)
+
+    if matches:
+        return purl_from_pattern("github", regex, uri)
+
+    segments = base_parsing(uri)
+
+    if segments:
+        name = segments[1]
+        namespace = segments[0]
+        version = None
+        subpath = None
+
+        # https://github.com/TG1999/fetchcode/master
+        if len(segments) >= 3 and segments[2] != "tree":
+            version = segments[2]
+            subpath = "/".join(segments[3:])
+
+        # https://github.com/TG1999/fetchcode/tree/master
+        if len(segments) >= 4 and segments[2] == "tree":
+            version = segments[3]
+            subpath = "/".join(segments[4:])
+
+        return PackageURL(
+            type="github",
+            namespace=namespace,
+            name=name,
+            version=version,
+            subpath=subpath,
+        )
+    return
+
+
+@purl_router.route("https?://bitbucket\\.org/.*")
+def build_bitbucket_purl(uri):
+    """
+    Take `uri` as input
+    For exmaple:
+    https://bitbucket.org/TG1999/first_repo/src/master
+    """
+    segments = base_parsing(uri)
+
+    if segments:
+        name = segments[1]
+        namespace = segments[0]
+        version = None
+        subpath = None
+
+        # https://bitbucket.org/TG1999/first_repo/new_folder/
+        if len(segments) >= 3 and segments[2] != "src":
+            version = segments[2]
+            subpath = "/".join(segments[3:])
+
+        # https://bitbucket.org/TG1999/first_repo/src/master/new_folder/
+        if len(segments) >= 4 and segments[2] == "src":
+            version = segments[3]
+            subpath = "/".join(segments[4:])
+
+        return PackageURL(
+            type="bitbucket",
+            namespace=namespace,
+            name=name,
+            version=version,
+            subpath=subpath,
+        )
+    return
+
+
+@purl_router.route("https?://gitlab\\.com/.*")
+def build_gitlab_purl(uri):
+    """
+    Take `uri` as input
+    For example:
+    https://gitlab.com/TG1999/firebase/-/tree/1a122122/views
+    """
```

Comment:
Again, we want docstrings to be more descriptive.

Reply:
Sure :)